### ISSUE TITLE
tensorflow couldn't use nvidia GPU in v1.7_cuda-12.3_ubuntu-22.04

### ISSUE THREAD

**Issue by a90091343** [iot-salzburg/gpu-jupyter#147]
# GPU-Jupyter Issue Report

## Issue Description

**Describe the issue**
In v1.7_cuda-12.3_ubuntu-22.04, pytorch can use GPU, but tensorflow 2.16.1 couldn't use the nvidia GPU.

**Tensorflow Output**

2024-06-12 14:33:35.816093: W tensorflow/core/common_runtime/gpu/gpu_device.cc:2251] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...

## Environment

**Operating System:**
Ubuntu 22.04.3. docker and nvidia-container-toolkit are all correctly installed and set.

**NVIDIA GPU and CUDA version Details:**

nvidia-smi
`NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4    `

nvcc --version
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Wed_Nov_22_10:17:15_PST_2023
Cuda compilation tools, release 12.3, V12.3.107
Build cuda_12.3.r12.3/compiler.33567101_0
```

**GPU-Jupyter Version:**
v1.7_cuda-12.3_ubuntu-22.04_python-only

## Probable Solution

I noticed that TensorFlow 2.16 only supports cuDNN 8.9 (https://tensorflow.google.cn/install/source?hl=en#gpu), while the `v1.7_cuda-12.3_ubuntu-22.04` Docker Image uses `FROM nvidia/cuda:12.3.2-cudnn9-runtime-ubuntu22.04`, which means there is a conflict in cuDNN versions. So, I changed the above line in `custom/header.Dockerfile` to:

```
FROM nvidia/cuda:12.3.2-runtime-ubuntu22.04
# ***** cudnn & tensorrt *****
ENV NV_CUDNN_VERSION 8.9.7.29
ENV NV_CUDNN_PACKAGE_NAME "libcudnn8"

ENV NV_CUDNN_PACKAGE "\${NV_CUDNN_PACKAGE_NAME}=\${NV_CUDNN_VERSION}-1+cuda12.2"

ENV NV_TENSORRT_VERSION 8.6.1.6

LABEL maintainer "NVIDIA CORPORATION <cudatools@nvidia.com>"
LABEL com.nvidia.cudnn.version="\${NV_CUDNN_VERSION}"

RUN apt-get update && apt-get install -y --no-install-recommends \
    \${NV_CUDNN_PACKAGE} \
    && apt-mark hold \${NV_CUDNN_PACKAGE_NAME} \
    && apt-get install -y --no-install-recommends \
    libnvinfer8=8.6.1.6-1+cuda12.0 libnvinfer-plugin8=8.6.1.6-1+cuda12.0 \
    && apt-mark hold libnvinfer8 libnvinfer-plugin8 \
    && rm -rf /var/lib/apt/lists/*
# ***** cudnn & tensorrt *****
```

After my testing, this solution resolved the issue of TensorFlow not being able to use the GPU and can now call CUDA correctly.

---

**Comment by benz0li** [comment]
@a90091343 You may be interested in b-data's/my [CUDA-based JupyterLab Python docker stack](https://github.com/b-data/jupyterlab-python-docker-stack/blob/main/CUDA.md).

ℹ️ See the [CUDA Version Matrix](https://github.com/b-data/jupyterlab-python-docker-stack/blob/main/CUDA_VERSION_MATRIX.md) for available Python versions (tags), PyTorch/TensorFlow compatibility, Recommended NVIDIA driver (Regular) and Supported NVIDIA drivers (LTSB).

**Comment by benz0li** [comment]
Cross reference regarding cuDNN versions in [`nvidia/cuda`](https://hub.docker.com/r/nvidia/cuda) images:

> Our current system is designed to build CUDA images with latest compatible cuDNN available and we expect future 12.4.x images to include cuDNN9 or higher.

– [CUDA 12.4 Container Images (#225) · Issues · nvidia / container-images / cuda · GitLab](https://gitlab.com/nvidia/container-images/cuda/-/issues/225#note_1844749131)

**Comment by ChristophSchranz** [comment]
Hi @a90091343 ,
I could not reproduce your issue on my setup. I noted that TF 2.16 does not officially support CUDA 12.3, seemed to work. To fix this issue, I've downgraded to CUDA 12.2 with cuDNN 8.

Please let me know if the setup in branch https://github.com/iot-salzburg/gpu-jupyter/tree/v1.7_cuda-12.2_ubuntu-22.04 solves your issue! The appropriate images will be available at [https://hub.docker.com/r/cschranz/gpu-jupyter/tags](https://hub.docker.com/r/cschranz/gpu-jupyter/tags?page=&page_size=&ordering=&name=v1.7_cuda-12.2) soon.

**Comment by a90091343** [comment]
Hi @ChristophSchranz ,
I have used the v1.7_cuda-12.2_ubuntu-22.04 image. In this image, I can use GPU to accelerate my TensorFlow codes. Thanks!
